### PR TITLE
A class can be declared in a lambda expression.

### PR DESCRIPTION
**Pull request by GerardPaligot**
Closes https://github.com/INRIA/spoon/issues/507

### CODE REVIEW

File: src/main/java/spoon/support/reflect/reference/CtTypeReferenceImpl.java [hunk 1]
```diff
@@ -157,23 +157,7 @@ protected AnnotatedElement getActualAnnotatedElement() {
...
-					return getQualifiedName().equals(element.getAnonymousClass().getQualifiedName());
-				}
-			});
-			if (elements.size()  == 0) {
-				return null;
-			}
-			return (CtType<T>) elements.get(0).getAnonymousClass();
-		}
-		return null;
+		return getFactory().Type().get(getQualifiedName());
```
Comment (monperrus) [inline src/main/java/spoon/support/reflect/reference/CtTypeReferenceImpl.java]: +1

File: src/main/java/spoon/support/reflect/reference/CtTypeReferenceImpl.java [hunk 2]
```diff
@@ -479,6 +463,13 @@ public boolean isAnonymous() {
 	}
 
 	@Override
+	public boolean isEnclosingType() {
+		final Pattern pattern = Pattern.compile("^([0-9]+)([a-zA-Z]+)$");
```
Comment (monperrus) [inline src/main/java/spoon/support/reflect/reference/CtTypeReferenceImpl.java]: for sake of future maintenance, explain what this means?

add a pointer to a document stating this convention for compilers? (or at least for JDT)?

Comment (GerardPaligot) [inline src/main/java/spoon/support/reflect/reference/CtTypeReferenceImpl.java]: > for sake of future maintenance, explain what this means?

Do you want a comment which explains that the name of a local type starts with a number and end with a letter?

> add a pointer to a document stating this convention for compilers? (or at least for JDT)?

There is no pointer for that. It doesn't exist a fully qualified name for local types ([Oracle doc](https://docs.oracle.com/javase/specs/jls/se7/html/jls-6.html#jls-6.7) and [JDT doc](http://help.eclipse.org/juno/index.jsp?topic=%2Forg.eclipse.jdt.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fjdt%2Fcore%2Fdom%2FITypeBinding.html)) but both have an identifier to identify the right local type.

File: src/main/java/spoon/reflect/declaration/CtTypeInformation.java [hunk 3]
```diff
@@ -60,6 +60,11 @@
 	boolean isAnonymous();
 
 	/**
+	 * Return {@code true} if the referenced type is declared in a executable.
+	 */
+	boolean isEnclosingType();
```
Comment (monperrus) [inline src/main/java/spoon/reflect/declaration/CtTypeInformation.java]: brainstorm for a better name:

* isInExecutable
* isLocalOrAnonymousClass (as in java.lang.Class)
* isLocalOrAnonymousOrLambdaClass (are all lambdas compiled to classes? )

Comment (GerardPaligot) [inline src/main/java/spoon/reflect/declaration/CtTypeInformation.java]: I love this one: `isLocalOrAnonymousClass` (as in java.lang.Class)

But it should be `isLocal`. We already have `isAnonymous`.

Comment (monperrus) [inline src/main/java/spoon/reflect/declaration/CtTypeInformation.java]: Then isLocalClass (as in java.lang.Class).

and we can write a nice contract:

for all klass:

  compile(klass).isLocalClass == ctclass(klass).isLocalClass ==
ctclass(klass).reference.isLocalClass

**Comment by monperrus** [comment]
Indeed:
```
 Object o = new Object(){
	 @Override
	 public String toString() {return "foo";};
 };

 System.out.println(o.getClass().getSimpleName());
 System.out.println(o.getClass().getCanonicalName());
 System.out.println(o.getClass().getName());
 System.out.println(o.getClass().isAnonymousClass());
```
returns 
```

null
Foo$1
true
```

```
 class Titi {}
 System.out.println(Titi.class.getSimpleName());
 System.out.println(Titi.class.getCanonicalName());
 System.out.println(Titi.class.getName());
 System.out.println(Titi.class.isAnonymousClass());
```
returns 
```
Titi
null
Foo$1Titi
false
```

File: src/main/java/spoon/reflect/declaration/CtTypeInformation.java [hunk 4]
```diff
@@ -60,6 +60,11 @@
 	boolean isAnonymous();
 
 	/**
+	 * Return {@code true} if the referenced type is declared in an executable.
+	 */
+	boolean isLocal();
```
Comment (monperrus) [inline src/main/java/spoon/reflect/declaration/CtTypeInformation.java]: not directly specified in any assert

File: src/main/java/spoon/support/reflect/declaration/CtTypeImpl.java [hunk 5]
```diff
@@ -345,6 +345,11 @@ public boolean isAnonymous() {
 	}
 
 	@Override
+	public boolean isLocal() {
+		return getReference().isLocal();
```
Comment (monperrus) [inline src/main/java/spoon/support/reflect/declaration/CtTypeImpl.java]: here we can do better than the naming convention: `getParent(CtMethod)!=null && getParent()!=CtNewClass` 

correct?

File: src/main/java/spoon/reflect/declaration/CtType.java [hunk 6]
```diff
@@ -34,6 +34,13 @@
 	String INNERTTYPE_SEPARATOR = "$";
 
 	/**
+	 * Returns the simple (unqualified) name of this element. If the type is
+	 * a local type, the name starts with its identifier.
+	 */
```
Comment (monperrus) [inline src/main/java/spoon/reflect/declaration/CtType.java]: Returns the simple (unqualified) name of this element. Following the compilation convention, if the type is a local type, the name starts with a numeric prefix (e.g. local class Foo has simple name 1Foo).

File: src/main/java/spoon/reflect/declaration/CtTypeInformation.java [hunk 7]
```diff
@@ -60,6 +60,25 @@
 	boolean isAnonymous();
 
 	/**
+	 * Return {@code true} if the referenced type is declared in an executable.
+	 * e.g. a type declared in a method or a lambda.
+	 *
```
Comment (monperrus) [inline src/main/java/spoon/reflect/declaration/CtTypeInformation.java]: This corresponds to `isLocalClass` of java.lang.Class.

File: src/main/java/spoon/reflect/reference/CtTypeReference.java [hunk 8]
```diff
@@ -34,6 +34,13 @@
 	String NULL_TYPE_NAME = "<nulltype>";
 
 	/**
+	 * Gets the simple name of referenced element. If the type reference is
+	 * a local type, the name starts with its identifier.
```
Comment (monperrus) [inline src/main/java/spoon/reflect/reference/CtTypeReference.java]: see other comment

**Comment by GerardPaligot** [comment]
@monperrus PR ok.

**Pull request merged**